presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoResultSet.java

```diff
@@ -1842,16 +1845,15 @@ public void interrupt(InterruptedException e)
         @Override
         public boolean hasNext()
         {
-            if (!rowQueue.isEmpty()) {
-                return true;
+            try {
+                semaphore.acquire();
```

Comment:
That is not correct from `Iterator`'s API perspective.

`hasNext()` can be invoked once, multiple times or not at all.
So, calling `hasNext()` must have no visible side effects.
(It can still change the internal state as it is eg in `com.google.common.base.AbstractIterator`).

BTW using that class could make the code simpler.
Then perhaps all you would need is blocking on `rowQueue`, without a need for a semaphore?

Reply:
You are right, and that *is* one place where AbstractIterator helps.
I do not think I can get rid of the Semaphore, though. (See comments above, about the case when the queue is full. In that case the Future couldn't finish, without failing to indicate to the consumer that it is done.)

Reply:
Right.
This is where Either would be a fit, to represent a poison pill
In Java terms, this could be Optional<Row> with empty being the pill.

Admittedly not beautiful, but simplifies the flow.
wdyt?

Reply:
The problem is that you cannot always place the poison pill on the queue.

I actually had a version with this, but I realized it would be tricky to get right when I added the code placing the poison pill (I called it Sentinel, but it's the same) to the Future, because that now might throw an InterruptedException, I couldn't figure out what to do in that case (loop until you can place the pill? give up?) - without busy waiting for two conditions at least.

Or do you mean some "outside" poison pill (not placed on the queue)? In that case it'd be necessary to have synchronization for this as well.

Reply:
@lhofhansl thanks. We could assume the `InterruptedException` won't fly when queue is not full and then adding a `volatile hasFailed` could perhaps be sufficient technically, but this wouldn't make the code look pretty.

Reply:
You'd have to somehow wait for that volatile to be changed *and* the queue. :)

(I can't believe how tedious this is, and Java - and Guava - have no pattern for this readily available. All you need is a BlockingQueue that you can poke upon ending - something like a close() method that would unblock a waiting take() call or something.)